include/freerdp/utils/chunkstream.h

```diff
@@ -0,0 +1,215 @@
+	 *
+	 * chunkstream_sizeAfterSlot is useful to compute the size of chunks after a certain slot,
+	 * typically when you want to write some headers that inform of the size of the following
+	 * payload.
+	 *
+	 * A full example usage with a protocol with at least 2 layers:
+	 * 		void layer2(ChunkStream *cs)
+	 * 		{
+	 * 			... allocate and write some more chunks ...
+	 * 		}
+	 *
+	 * 		void layer1(ChunkStream *cs)
+	 * 		{
+	 * 			wStream layer1Headers;
+	 * 			// headers of layers 1 are at max 20 bytes
+	 * 			ChunkStreamSlot *headersSlot = chunkstream_getPoolStream(cs, 20, &layer1Headers);
+	 *
+	 * 			layers2(cs); // create payload, will adds chunks in the chunkstream
+	 *
+	 * 			int payloadSz = chunkstream_sizeAfterSlot(cs, headersSlot);
+	 *
+	 * 			writeLayer1Headers(layer1Headers, payloadSz);
+	 *
+	 * 			chunkstreamslot_update_fromStream(headersSlot, &layer1Headers);
+	 * 		}
+	 *
+	 * 		void writePacket()
+	 * 		{
+	 * 			ChunkStream *cs = chunkstream_new(1024);
+	 * 			layer1(cs);
+	 *
+	 * 			wStream *s = chunkstream_linearizeToStream(cs);
+	 *
+	 * 			... serialize the content of s ...
+	 * 		}
+	 */
+
+	/** Initialises a chunkstream
+	 * @param initialSize the initial capacity of the chunkstream's pool
+	 * @return if the created chunkstream
+	 */
+	FREERDP_API ChunkStream* chunkstream_new(size_t initialSize);
+
+	/** Destroys internal data used by this chunkstream
+	 * @param pchunkstream a pointer to a chunkstream
+	 */
+	FREERDP_API void chunkstream_destroy(ChunkStream** pchunkstream);
+
+	/**
+	 * Retrieves a slot containing a static string
+	 * @param chunkstream the chunkstream
+	 * @param str the string to add
+	 * @param include0 tells if we should include the leading null byte
+	 * @return the corresponding ChunkStreamSlot, NULL if it failed (no available slots)
+	 */
+	FREERDP_API ChunkStreamSlot* chunkstream_getStaticStringSlot(ChunkStream* chunkstream,
+	                                                             const char* str, BOOL include0);
+
+	/**
+	 * Retrieves a slot containing a static memory blob
+	 * @param chunkstream the chunkstream
+	 * @param data the memory blob to add
+	 * @param sz size of the memory blob
+	 * @return the corresponding ChunkStreamSlot, NULL if it failed (no available slots)
+	 */
+	FREERDP_API ChunkStreamSlot* chunkstream_getStaticMemSlot(ChunkStream* chunkstream,
+	                                                          const BYTE* data, size_t sz);
+
+	/**
+	 * Retrieves a slot containing a memory blob alloced by the user, the chunkstream takes
+	 * ownership of the provided memory, so it will be freed when the chunkstream is destroyed.
+	 * @param chunkstream the chunkstream
+	 * @param data the memory blob to add
+	 * @param sz size of the memory blob
+	 * @return the corresponding ChunkStreamSlot, NULL if it failed (no available slots)
+	 */
+	FREERDP_API ChunkStreamSlot* chunkstream_getMallocSlot(ChunkStream* chunkstream, BYTE* data,
```

Comment:
so, you assume the memory is allocated with `malloc`? (there are also the `aligned_malloc` et al which require a custom deallocator)

Reply:
Hum, `ChunkStream` is sending packets oriented, so I don't know if we're really using memory allocated by `aligned_malloc` when forging packets, but I can add a function in that sense if needed.